src/update_flash.c

```diff
@@ -492,6 +493,19 @@ static int RAMFUNCTION wolfBoot_update(int fallback_allowed)
                 wolfBoot_set_update_sector_flag(sector, flag);
         }
         sector++;
+
+        /* headers that can be in different positions depending on when the
+         * power fails are now in a known state, re-read and swap fw_size
+         * because the locations are correct but the metadata is now swapped */
+        if (sector == 1) {
+            wolfBoot_open_image(&boot, PART_BOOT);
```

Comment:
I think this is an incomplete fix:
If a power failure happens when writing the header to the boot partition, the status will be:
- the update header is overwritten with the backup of the boot header
- the swap sector contains the update header
- the boot header is erased/corrupted
In this case, the `total_size` will be calculated using only the v1 image size (header in the update partition). The image v2 size is "hidden" in the swap sector and it isn't checked.
To trigger the use case in a test:
1. modify the size of the v2 image:
```
--- a/tools/test.mk
+++ b/tools/test.mk
@@ -173,7 +173,7 @@ test-sim-internal-flash-with-update: wolfboot.bin test-app/image.elf FORCE
        $(Q)dd if=/dev/urandom of=test-app/image.elf bs=1K count=16 oflag=append conv=notrunc
        $(Q)$(SIGN_TOOL) $(SIGN_OPTIONS) test-app/image.elf $(PRIVATE_KEY) 1
        $(Q)cp test-app/image.bak.elf test-app/image.elf
-       $(Q)dd if=/dev/urandom of=test-app/image.elf bs=1K count=16 oflag=append conv=notrunc
+       $(Q)dd if=/dev/urandom of=test-app/image.elf bs=1K count=32 oflag=append conv=notrunc
        $(Q)$(SIGN_TOOL) $(SIGN_OPTIONS) test-app/image.elf $(PRIVATE_KEY) $(TEST_UPDATE_VERSION)
        $(Q)dd if=/dev/zero bs=$$(($(WOLFBOOT_SECTOR_SIZE))) count=1 2>/dev/null | tr "\000" "\377" > erased_sec.dd
        $(Q)$(SIGN_TOOL) $(SIGN_ARGS) $(DELTA_UPDATE_OPTIONS) \
```
2. test with a *single* failure update when writing to the boot header:
```
# prints 1
./wolfboot.elf update_trigger get_version 
./wolfboot.elf powerfail c0020000
# prints 1 (should print 2)
./wolfboot.elf get_version 
```

Please add this edge case to the tests when fixed

Reply:
good point, didn't think about total_size